### PR TITLE
Do not store webview if there is no corresponding serializer

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

This PR is a duplicate of previously closed https://github.com/eclipse-theia/theia/pull/8473

#### What it does

- Disables saving of the webview state if the plugin don't register an apropriate theia.WebviewPanelSerializer.
- Fixes an error with restoring the editor area when a webview and several editors are opened.

Video demonstrating bug https://www.youtube.com/watch?v=9ir1hTov350

Fixes https://github.com/eclipse-theia/theia/issues/8620 and https://github.com/eclipse-theia/theia/issues/6877

This is another approach in comparing with https://github.com/eclipse-theia/theia/pull/8621
In #8621 Theia stores the empty state for the webview and tries to find the serializer when restoring the state.
In this PR, Theia checks for the theia.WebviewPanelSerializer, and does not store the webview  to the layout at all if the serializer is not found.

#### How to test

To test it you need to have a plugin which opens a webview, but don't register `vscode.WebviewPanelSerializer` for this webview.
I used this one https://github.com/vitaliy-guliy/greeting-extension, which opens a Greeting webview each time when Theia starts.

Steps to reproduce:
- run Theia with the plugin, open the browser. Greeting plugin will open the webview after startup.
- open few files, Greeting tab must be stay first
- reload the browser

Opened flies must be restored, the order of files must be saved. As the webview is not persisted, after reloading it will be opened again next to active tab.

Follow 'Steps to reproduce' in the original issue https://github.com/eclipse-theia/theia/issues/8620

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by vitaliy-guliy** [comment]
@akosyakov what do you think about not to store the webview at all in a case when there is no webview serializer. Instead of saving it with `{}` inner state and then disposing the widget when restoring the layout.

**Review by tsmaeder — COMMENTED** [review]
I think we should handle the case where widgets cannot be recreated. For example, what if a plugin is updated and no longer contributes a webview serializer? 
`ShellLayoutRestorer#convertToWidget` can return `undefined` and we need to handle that.

**Comment by vitaliy-guliy** [comment]
> I think we should handle the case where widgets cannot be recreated. For example, what if a plugin is updated and no longer contributes a webview serializer?
> `ShellLayoutRestorer#convertToWidget` can return `undefined` and we need to handle that.

Agree with you. I think thew we need to handle both cases:
- do not store the webview state if there is no serializer
- handle `undefined` when converting JSON to widget

**Comment by vitaliy-guliy** [comment]
@tsmaeder done. I've handled a case when the widget cannot be recreated.

File: packages/core/src/browser/shell/application-shell.ts
Comment (tsmaeder) [inline packages/core/src/browser/shell/application-shell.ts:639]: T.b.h I liked the previous version better: can't we just filter the array in `ShellLayoutRestorer#parse()` after we've created all the widgets?

Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts:639]: And I like the old version better because the filtering is done closer to where the widget is created (or not created). Simpler to understand when looking at ShellLayoutRestorere.

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: This what I wanted to do first, but in this case it will not work.
In `ShellLayoutRestorer#parse()` each widget is created in a separate promise, which is called already after returning array of widgets.

https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L272-L279

Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts:639]: We wait can for all promises to be fullfilled: we should not return before the `widgets` array is fully populated anyway:

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: In previous version we had a problem with order of opened editors. It's because the time for restoring/creating widgets for different files is different and the created widget was always inserted the first.

Layout data is created in `ShellLayoutRestorer#inflate` function. 
At the beginning of this function we parse layout string representation to have `LayoutData`.
https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L228

But at that moment arrays of widgets are empty for all panels.

At the end we call `parseContext.inflate(context)`. It runs all the promises, which create the widgets and insert them in the array.

https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L256

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: Everything goes in  `ShellLayoutRestorer#inflate`

The first we parse the layout, but without creating the widget instances.
Then we check for the layout version and perform some migrations.
And only then if everything's fine we run `parseContext.inflate(context)` and create the widgets.
Here the idea was do not create widgets before checks.

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: That's why to preserve order of opened editors we insert the widget to the array by index.

https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L276

**Review by azatsarynnyy — APPROVED** [review]
I've tested both use-cases on the provided plug-in https://github.com/vitaliy-guliy/greeting-extension:
- when a contributed WebView has a related registered serializer
- when a contributed WebView without a registered serializer

All widgets restore works well. In both use cases, the behavior is the same as in VS Code :+1: 
So, I'm +1 for this patch.

It would be great to have another pair of eyes look at it as the proposed changes affect one of the essential part - the application shell.
cc @eclipse-theia/core

File: packages/core/src/browser/shell/application-shell.ts
Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts:639]: Yes, but we can do 
```
for (let i = 0; i < descs.length; i++) {
  const promises= [];
  parseContext.push(async context => {
     promises.push(this.convertToWidget(descs[i], context).then(widget => widgets[i] = widget));
});}
await Promise.all(promises);
return widgets.filter(widget => widget !== undefined);
```

Comment (spoenemann) [reply packages/core/src/browser/shell/application-shell.ts:639]: Yes, it would be great to do this filtering directly where the arrays are filled, not where they are consumed.

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: @tsmaeder @spoenemann
If we create a widgets in the `ShellLayoutRestorer#parse` method by adding `await Promise.all(promises)`, then the code in `ShellLayoutRestorer#inflate` does not have a big sense.

The idea was:
- prepare `ShellLayoutRestorer.ParseContext` (array of promises)
- parse the layout without creating the widgets, but add promises to `ParseContext` for each widget
- perform some checks and migrations
- create widgets with `await parseContext.inflate(context)`
https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L256

Yes, it will be much clear if we do everything when parsing the JSON in `ShellLayoutRestorer#parse`. But again, then we need to completely rework `ShellLayoutRestorer#inflate`. 

@akosyakov would you like to comment? I see you introduced this code a year ago
https://github.com/eclipse-theia/theia/commit/f8a333ff025483c10fe57a79ed83a5d5173b888c#diff-5974c221773ca89483f3da1843aa00038cf2d6b0eb398fce80d3f31434f2e1ebR212

Comment (spoenemann) [reply packages/core/src/browser/shell/application-shell.ts:639]: Would it work to move the filtering code to the end of `ShellLayoutRestorer#inflate`, just before returning the layout data?

Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts:639]: Why do we need to change ShellLayouRestorer#parse if no null values are ever returned? Just remove that change!

Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts:639]: I thought we had a change originally that worked, it just did not respect the order. Return to that change and fix the ordering.

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: > Would it work to move the filtering code to the end of `ShellLayoutRestorer#inflate`, just before returning the layout data?

I think it will work, but the check will be only for specific panel
```
    if (layout.mainPanel && layout.mainPanel.main && layout.mainPanel.main.type === 'tab-area') {
        const config = layout.mainPanel.main as DockLayout.ITabAreaConfig;
        config.widgets = config.widgets.filter(value => value !== undefined);
    }
```

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: > I thought we had a change originally that worked, it just did not respect the order. Return to that change and fix the ordering.

Do you mean a change in https://github.com/eclipse-theia/theia/pull/8621 ?
At the first glance it looks nice. But the problem with ordering was because the time for restoring different editors with different type of files was different. There is no guarantee that promises that create widgets will be resolved in predefined sequence.
I wrote about that in my comment
https://github.com/eclipse-theia/theia/pull/8621#issuecomment-717889244

Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts:639]: Yes, that change. The code I proposed in https://github.com/eclipse-theia/theia/pull/8680/files#r520333442 takes care of the ordering problem.

Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts:639]: The "not storing the state if there is not serializer" still makes sense, though.

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: > Yes, that change. The code I proposed in https://github.com/eclipse-theia/theia/pull/8680/files#r520333442 takes care of the ordering problem.

Yes, it takes. But then widgets will be created before checks and migrations, that are in `ShellLayoutRestorer#inflate`

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: @tsmaeder @spoenemann I have another idea is to move filtering at the end of `ParseContext#inflate`.

```
    export class ParseContext {
        protected readonly toInflate: Inflate[] = [];
        protected readonly arrays: WidgetArray[] = [];

        newArray(): WidgetArray {
            const array: WidgetArray = [];
            this.arrays.push(array);
            return array;
        }

        async inflate(context: InflateContext): Promise<void> {
            const pending: Promise<void>[] = [];
            while (this.toInflate.length) {
                pending.push(this.toInflate.pop()!(context));
            }
            await Promise.all(pending);

            this.arrays.forEach(array => array = array.filter(value => value !== undefined));
        }
    }
```

Then `ShellLayoutRestorer#parse` can ask for a widget array, which will be filtered after creating the widgets
```
    const widgets = parseContext.newArray();
    const descs = (value as WidgetDescription[]);
    for (let i = 0; i < descs.length; i++) {
        parseContext.push(async context => {
            widgets[i] = await this.convertToWidget(descs[i], context);
        });
    }
    return widgets;
```

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts:639]: @tsmaeder @spoenemann I have another idea is to move filtering at the end of `ShellLayoutRestorer.ParseContext#inflate`.

But the `ParseContext` must have all the arrays to be filtered.

```
    export class ParseContext {
        protected readonly toInflate: Inflate[] = [];
        protected readonly toFilter: Widgets[] = [];

        newArray(): Widgets {
            const array: Widgets = [];
            this.toFilter.push(array);
            return array;
        }

        async inflate(context: InflateContext): Promise<void> {
            const pending: Promise<void>[] = [];
            while (this.toInflate.length) {
                pending.push(this.toInflate.pop()!(context));
            }
            await Promise.all(pending);

            if (this.toFilter.length) {
                this.toFilter.forEach(array => {
                    const filtered = array.filter(value => value !== undefined);
                    array.splice(0, array.length, ...filtered);
                });
            }
        }
    }

    export type WidgetArray = (Widget | undefined)[];
```

Then `ShellLayoutRestorer#parse` can ask for a widget array, which will be filtered after creating the widgets
```
    const widgets = parseContext.newArray();
    const descs = (value as WidgetDescription[]);
    for (let i = 0; i < descs.length; i++) {
        parseContext.push(async context => {
            widgets[i] = await this.convertToWidget(descs[i], context);
        });
    }
    return widgets;
```

I have just checked this approach, it works well.

Comment (spoenemann) [reply packages/core/src/browser/shell/application-shell.ts:639]: Sounds good.

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts]: Reworked

File: packages/core/src/browser/shell/shell-layout-restorer.ts
Comment (spoenemann) [inline packages/core/src/browser/shell/shell-layout-restorer.ts:385]: Performance: Do this `splice` only if `filtered.length` is smaller than `array.length`

File: packages/core/src/browser/shell/application-shell.ts
Comment (tsmaeder) [reply packages/core/src/browser/shell/application-shell.ts]: I really don't understand why this needs to be so complicated: just wait for all the promises to resolve and then filter out the undefined values from the array in ShellLayoutRestorer#parse(). It's better code-quality-wise because it does the filtering closer to where the nulls are created and there is no if conditional expression around the filtering. Please convince me!

File: packages/core/src/browser/shell/shell-layout-restorer.ts
Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/shell-layout-restorer.ts:385]: > Performance: Do this `splice` only if `filtered.length` is smaller than `array.length`

Interesting, I didn't know about it before. In fact the array will not have so much items to work very slow.
But thinking about performance. To avoid checking for an array length, maybe it's better to reset the array at all and then push all the items..
```
array.length = 0;
array.push(...filtered);
```
WDYT?

File: packages/core/src/browser/shell/application-shell.ts
Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/application-shell.ts]: It's such confusing because promises for the widgets are created in one place and then resolved in another. 
Why do we need so?
It could happen that your Theia is updated or downgraded and the layout cannot be restored properly.
That requires to do some checks for the layout version and probably transform the layout or update the state of some widgets before creating their instances.
Just take a look at `ShellLayoutRestorer#inflate()`

File: packages/core/src/browser/shell/shell-layout-restorer.ts
Comment (spoenemann) [reply packages/core/src/browser/shell/shell-layout-restorer.ts:385]: That might be an alternative to `splice`, but checking `length` for sure has lower cost than resetting the array. And resetting is only necessary if the length has changed.

Another alternative would be this:
```typescript
for (let i = 0; i < array.length; i++) {
    if (array[i] === undefined) {
        array.splice(i--, 1);
    }
}
```
If there are only few undefined elements, this variant might be even better because it does not duplicate the array.

Comment (vitaliy-guliy) [reply packages/core/src/browser/shell/shell-layout-restorer.ts]: In 99% array will not contain undefined elements, so this alternative looks much better than rebuilding the array each time when restoring the layout.
Thanks for this.
I've updated the PR.

**Comment by vitaliy-guliy** [comment]
I'm going to merge this PR tomorrow if there are no objections.

**Pull request merged**
